src/qibo/quantum_info/entropies.py

```diff
@@ -0,0 +1,924 @@
+
+    if abs(purity(state) - 1.0) < PRECISION_TOL:
+        return 0.0
+
+    if alpha == 1.0:
+        return von_neumann_entropy(state, base=base, backend=backend)
+
+    return (1 / (1 - alpha)) * (np.trace(_matrix_power(state, alpha, backend)) - 1)
+
+
+def entanglement_entropy(
+    state,
+    bipartition,
+    base: float = 2,
+    check_hermitian: bool = False,
+    return_spectrum: bool = False,
+    backend=None,
+):
+    """Calculates the entanglement entropy :math:`S` of bipartition :math:`A`
+    of ``state`` :math:`\\rho`. This is given by
+
+    .. math::
+        S(\\rho_{A}) = -\\text{tr}(\\rho_{A} \\, \\log(\\rho_{A})) \\, ,
+
+    where :math:`\\rho_{A} = \\text{tr}_{B}(\\rho)` is the reduced density matrix calculated
+    by tracing out the ``bipartition`` :math:`B`.
+
+    Args:
+        state (ndarray): statevector or density matrix.
+        bipartition (list or tuple or ndarray): qubits in the subsystem to be traced out.
+        base (float, optional): the base of the log. Defaults to :math: `2`.
+        check_hermitian (bool, optional): if ``True``, checks if :math:`\\rho_{A}` is Hermitian.
+            If ``False``, it assumes ``state`` is Hermitian . Default: ``False``.
+        return_spectrum: if ``True``, returns ``entropy`` and eigenvalues of ``state``.
+            If ``False``, returns only ``entropy``. Default is ``False``.
+        backend (:class:`qibo.backends.abstract.Backend`, optional): backend to be used
+            in the execution. If ``None``, it uses
+            :class:`qibo.backends.GlobalBackend`. Defaults to ``None``.
+
+    Returns:
+        float: Entanglement entropy :math:`S` of ``state`` :math:`\\rho`.
+    """
+    if backend is None:  # pragma: no cover
+        backend = GlobalBackend()
+
+    if base <= 0.0:
+        raise_error(ValueError, "log base must be non-negative.")
+
+    if (
+        (len(state.shape) not in [1, 2])
+        or (len(state) == 0)
+        or (len(state.shape) == 2 and state.shape[0] != state.shape[1])
+    ):
+        raise_error(
+            TypeError,
+            f"state must have dims either (k,) or (k,k), but have dims {state.shape}.",
+        )
+
+    nqubits = int(np.log2(state.shape[0]))
+
+    reduced_density_matrix = (
+        backend.partial_trace(state, bipartition, nqubits)
+        if len(state.shape) == 1
+        else backend.partial_trace_density_matrix(state, bipartition, nqubits)
+    )
+
+    entropy_entanglement = von_neumann_entropy(
+        reduced_density_matrix,
+        base=base,
+        check_hermitian=check_hermitian,
+        return_spectrum=return_spectrum,
+        backend=backend,
+    )
+
+    return entropy_entanglement
+
+
+def _matrix_power(matrix, alpha, backend):
+    """Calculates ``matrix ** alpha`` according to backend."""
+    if backend.__class__.__name__ in [
+        "CupyBackend",
+        "CuQuantumBackend",
+    ]:  # pragma: no cover
+        new_matrix = backend.to_numpy(matrix)
```

Comment:
isn't there a way to do this on the gpu?

Reply:
`cupy` does not have an equivalent to `scipy.linalg.fractional_matrix_power`. Maybe there is a way, but then another function should be used.